Implement MarkdownProjectReader

aunger opened the pull request:
For https://github.com/WycliffeAssociates/otter-jvm/issues/374

aunger committed:
Make a generic version of Tree.

aunger committed:
Implement MarkdownProjectReader.

aunger committed:
Skip blank help bodies.

KJoslyn commented on src/main/kotlin/org/wycliffeassociates/otter/common/domain/resourcecontainer/project/markdown/MarkdownProjectReader.kt:
```diff
@@ -1,13 +1,110 @@
+            .onLeave { treeRoot = treeCursor.pop() }
+            .filter { it.isFile }
+            .map { OtterTreeNode(it) }
+            .forEach { treeCursor.peek()?.addChild(it) }
```
This solution is very intriguing to me! Certainly a different way of thinking than typical sequential operations.

aunger replied:
Yeah, I'm not super proud of my code readability here. It could be a bit cleaner if our OtterTree nodes had parent links, but my treeCursor stack serves the purpose.

jsarabia commented on src/test/kotlin/org/wycliffeassociates/otter/common/domain/resourcecontainer/project/markdown/MarkdownProjectReaderTest.kt:
```diff
@@ -0,0 +1,69 @@
+import org.wycliffeassociates.otter.common.collections.tree.OtterTreeNode
+import org.wycliffeassociates.otter.common.domain.resourcecontainer.project.IProjectReader
+import java.io.File
+import java.nio.file.Paths
```
I was actually surprised and since I last tried to use java.nio, some of it has actually been added to Android. However, it was introduced in 8.0; but we typically have to target 4.4 or 5.0.

Would you mind if we moved this test to the JVM repo?

aunger replied:
@sarabiaj I can move it, but would it be ok to instead remove this NIO import and change [line 12](https://github.com/WycliffeAssociates/otter-common/pull/87/files#diff-b8fa3e633ce32d9cc8e9ef3e6e274489R12) to this: 
```private val pwd = File(System.getProperty("user.dir"))```
You still won't want to *run* this test on Android, since under Android `user.dir=/` which would make these tests take forever (since they walk the whole directory tree), but I don't think we'd ever be executing unit tests on Android, right?

jsarabia replied:
I'm sure you wouldn't even get the file permissions to even be able to do it.

I suppose if we did get to porting to Android and did want to run tests, we'd probably be more concerned with more Android centric things; but I believe there are services to throw tests at a bunch of Android devices and get the results back lol.

Either is fine with me though.

aunger replied:
Removed NIO import.

aunger committed:
Remove java.nio import.

jsarabia commented on src/main/kotlin/org/wycliffeassociates/otter/common/domain/resourcecontainer/project/markdown/MarkdownProjectReader.kt:
```diff
@@ -1,13 +1,110 @@
+            Collection(
+                    sort = id,
+                    slug = slug,
+                    labelKey = "helpcollection",
```
this sounds like it would be too vague of a key?

aunger replied:
True.  Two answers though:
1. The specific (shortsighted) answer: It doesn't matter, because these collections aren't written to the DB in any of our use-cases; the content gets attached to the existing collection hierarchy from the already-imported Bible RC.
2. The general answer: This MarkdownProjectReader should not limit itself to existing use-cases, and so I should either parameterize this value, or use something from the manifest instead of this hardcoded value.    

I'll push an update.

aunger replied:
Done.  We'll eventually want this key to be more configurable (like when we start reading Bibles in Markdown format), and MarkdownProjectReader should be parametrized with a function to create the key from the RC+slug (or something) but I think we should defer that impl until we know more about what we'll need.  For now, it just uses `dublinCore::identifier`.

aunger committed:
Don't hardcode Markdown collection key.

We'll eventually want this key to be more configurable (like
when we start reading Bibles in Markdown format), and
MarkdownProjectReader should be parametrized with a function
to create the key from the RC+slug (or something) but I'm
deferring that impl until we know more about what we'll need.

aunger committed:
Merge branch 'dev' into rba-tn-md-project-tree
